File: 1st Place/src/models/model_weekday.R

# Please change the working directory to the parent folder before running

#### enviroment setup ####

require(zoo)
require(dplyr)
require(xgboost)

#### function defination ####
read_data = function(filename){
  
  # The function is used to read raw data and aggregate by hour
  # Parameter: filename -- path of data
  # Return: dataframe of hourly data

  df = read.csv(filename,stringsAsFactors = FALSE)
  df$Timestamp = strptime(df$Timestamp,format = '%Y-%m-%d %H:%M:%S')
  # truncate timestamp
  df$Timestamp = paste(substr(df$Timestamp,1,13),':00:00',sep = '')
  
  new_df = summarise(group_by(df,meter_id,Timestamp,is_abnormal),
                     Values = mean(Values,na.rm = TRUE))
  return(new_df)
}
read_weather = function(filename_weather){
  # The function is used to read weather data and aggregate by hour
  # Parameter: filename -- path of data
  # Return: dataframe of hourly data
  weather = read.csv(filename_weather,stringsAsFactors = FALSE)
  weather$hour = paste(substr(weather$Timestamp,1,13),':00:00',sep = '')
  
  new_weather = summarise(group_by(weather,hour,Distance,site_id),
                          Temperature = mean(Temperature,na.rm = TRUE))
  colnames(new_weather)[1] = 'Timestamp'
  
  return(new_weather)
}

extract_features = function(df,weather){
  
  # The function is used to extract features from input_data
  # The features include:'month', 'week', 'hour', 'dayofyear','dayofmonth',
  #                     'dayofweek', 'temperature', 'is_off'
  # Parameter: df -- input data of current meter
  # Parameter: weather -- input weather data of current meter
  # Return: dataframe with features stated above
  
  df$month = (as.POSIXlt(df$Timestamp))$mon
  df$week = ceiling(((as.POSIXlt(df$Timestamp))$yday)/7)
  df$hour = (as.POSIXlt(df$Timestamp))$hour
  df$dayofyear = (as.POSIXlt(df$Timestamp))$yday
  df$dayofmonth = (as.POSIXlt(df$Timestamp))$mday
  df$dayofweek = (as.POSIXlt(df$Timestamp))$wday
  
  tmp_weather = subset(weather,select = c('Timestamp','Temperature'))
  df = merge(df,tmp_weather,by = 'Timestamp',all.x = TRUE)
  df$is_off = (df$dayofweek == 0)|(df$dayofweek == 6)
#  df$is_pre_off = df$dayofweek <= 1 
#  df$is_next_off = df$dayofweek >= 5
  
  return(df)
}
handle_holidays = function(df,holidays){
  # The function is used to add holiday information,and will change 
  # is_off,is_pre_off,is_next_off is the corresponding day is a public holiday
  # param: df -- site data after feature extraction
  # param: holidays -- site public holidays
  # return: df -- modified site data
  if(length(nrow(holidays)) == 0){
    return(df)
  }
  for(i in 1:nrow(holidays)){
    tmp_date = as.Date(holidays$Date[i])
    df$is_off[which(as.Date(df$Timestamp) == tmp_date)] = 1
  }
  return(df)
}
holiday_correction = function(df){
  # The function label days with abnormal low value with holiday label(is_off = 1).
  # Also it add two features is_pre_off & is_next_off, which indicates whether the day before/after
  # is a "holiday/low-value day"
  # param: df -- site_data after feature_extraction
  # return: df -- modifed site_data
  tmp1 = summarise(group_by(df,is_off),mean = mean(Values),std = sd(Values))
  tmp2 = summarise(group_by(df,as.Date(df$Timestamp),is_off),mean = mean(Values))
  day_low = tmp2$`as.Date(df$Timestamp)`[which(tmp2$is_off == 0 & tmp2$mean <= 
                                                 (tmp1$mean[2] + tmp1$std[2]))]
  for(i in 1:length(day_low)){
    tmp_date = as.Date(day_low[i])
    df$is_off[which(as.Date(df$Timestamp) == tmp_date)] = 1
  }
  
  df$date = as.Date(df$Timestamp)
  tmp3 = summarise(group_by(df,date,is_off))
  tmp3 = tmp3[order(tmp3$date),]
  
  tmp3$is_pre_off = c(0,tmp3$is_off[1:(nrow(tmp3)-1)])
  tmp3$is_next_off = c(tmp3$is_off[2:nrow(tmp3)],0)
  tmp4 = subset(tmp3,select = -is_off)
  
  df = merge(df,tmp4, by = 'date',all.x = TRUE)
  return(df)
}

xgb_model = function(df,n_fold){
  # The function is used to perform "n-fold" xgboost.Please refer to the report for details.
  # param: df -- site data
  # param: n_fold -- the number of cycles
  # return: dataframe with prediction value(average and details), true value and average error.
  
  features = c('month', 
               #'week',
               'hour', 'dayofyear',
              'dayofmonth', 'dayofweek', 'Temperature', 'is_off', 'is_pre_off',
              'is_next_off')
  X = data.matrix(subset(df,select = features))
  Y = data.matrix(df$Values)
  cut_number = cut(seq_len(nrow(X)),breaks = n_fold,labels = FALSE)
  
  result = data.frame(Timestamp = df$Timestamp,value_true = Y)
  for(i in 1:n_fold){
    X_train = X[-which(cut_number == i),]
    X_test = X[which(cut_number == i),]
    Y_train = Y[-which(cut_number == i)]
    Y_test = Y[which(cut_number == i)]
    
    dtrain = xgb.DMatrix(data = X_train,label = Y_train)
    dtest = xgb.DMatrix(data = X_test,label = Y_test)
    
    watchlist = list(train = dtrain,test = dtest)
    bst = xgb.train(data = dtrain,nround = 10000,watchlist = watchlist,
                    early_stopping_rounds = 100,
                    objective = 'reg:linear')
    pre = predict(bst,X)
    
    result[,paste('pre_',i,sep = '')] = pre
  }
  
  result$pre_ave = apply(result[,3:(2+n_fold)],1,mean)
  result$err = result$value_true - result$pre_ave
  
  return(result)
}
get_anomalies = function(xgb_result,n_sigma){
  # This function is used to select anomalies from xgboost result
  # param: xgb_result -- xgboost result from xgb_model function
  # param: n_sigma -- use n_sigma rule to select anomalies
  # return: return a list containing abnormal days & xgboost result by day
  xgb_result$date = as.Date(xgb_result$Timestamp)
  result_day = summarise(group_by(xgb_result,date),
                         true = mean(value_true),pre = mean(pre_ave),
                         err = mean(err))
  err_mean = mean(result_day$err)
  err_std = sd(result_day$err)
  anomalies = result_day$date[which(result_day$err >= err_mean+n_sigma*err_std)]
  
  rst_list = list(result_day,anomalies)
  return(rst_list)
}
pipeline = function(df,weather,holidays,n_fold,n_sigma){
  
  print('Extract features')
  df = extract_features(df,weather)
  df = handle_holidays(df,holidays)
  df = holiday_correction(df)
  df = df[-which(is.na(df$Temperature)),]
  print('Use xgb to train model')
  xgb = xgb_model(df,n_fold)
  print('Select out anomalies')
  rst = get_anomalies(xgb,n_sigma)
  return(rst)
}

#### main code ####

# read in holidays & weather
holidays = read.csv('data/raw/holidays.csv',stringsAsFactors = FALSE)
weather = read_weather('data/raw/weather.csv')

# anomalies site 234
print("Working on site 234 ... ")
weather_234 = weather[which((weather$site_id == '234_203') 
                            & (weather$Distance >10) & (weather$Distance < 13)),]
holidays_234 = FALSE
df_234 = read_data('data/processed/234_203.csv')
df_234 = df_234[which(as.Date(df_234$Timestamp) <= as.Date('2015-06-20')),]
rst_234 = pipeline(df_234,weather_234,holidays_234,4,3)
anomalies_234 = data.frame(meter_id = '234_203',date = rst_234[[2]])

# anomalies site 334
print("Working on site 334 ... ")
weather_334 = weather[which((weather$site_id == '334_61') & (weather$Distance >= 19)),]
holidays_334 = holidays[which(holidays$site_id == '334_61'),]
df_334 = read_data('data/processed/334_61.csv')
rst_334 = pipeline(df_334,weather_334,holidays_334,4,4)
anomalies_334 = data.frame(meter_id = '334_61',date = rst_334[[2]])
print(anomalies_334)

# anomalies site 38
print("Working on site 38 ... ")
weather_38 = weather[which((weather$site_id == '38')),]
holidays_38 = holidays[which(holidays$site_id == '038'),]
df_38 = read_data('data/processed/38_9687.csv')

df_38_1 = df_38[which(as.Date(df_38$Timestamp) >= as.Date('2012-12-21')),]
rst_38_1 = pipeline(df_38_1,weather_38,holidays_38,4,3.5)
df_38_2 = df_38[which(as.Date(df_38$Timestamp) <= as.Date('2012-12-21')),]
rst_38_2 = pipeline(df_38_2,weather_38,holidays_38,4,3.5)

anomalies_38 = rbind(data.frame(meter_id = '38_9686',date = rst_38_1[[2]]),
                     data.frame(meter_id = '38_9686',date = rst_38_2[[2]]))

# write result
rst_weekday = rbind(rbind(anomalies_234,anomalies_334),anomalies_38)

raw_234 = read.csv('data/processed/234_203.csv',stringsAsFactors = FALSE)
raw_334 = read.csv('data/processed/334_61.csv',stringsAsFactors = FALSE)
raw_38 = read.csv('data/processed/38_9687.csv',stringsAsFactors = FALSE)
raw = rbind(rbind(raw_234,raw_334),raw_38)

for(i in 1:nrow(rst_weekday)){
  raw[which((raw$meter_id == rst_weekday$meter_id[i]) & (substr(raw$Timestamp,1,10) == rst_weekday$date[i])),
      'is_abnormal'] = 'True'
}
print(length(which(raw$is_abnormal == 'True')))
write.csv(raw[,1:4],'data/result/res_weekday_r.csv',row.names = FALSE,quote = FALSE)

File: 1st Place/src/models/model_weekday.py
# -*- coding: utf-8 -*-
import pandas as pd
import os
import xgboost as xgb
from datetime import datetime,timedelta

PARENT_FOLDER = os.path.abspath(os.path.join(__file__ ,"../../.."))
DATA_PATH = os.path.join(PARENT_FOLDER,'data/processed/')
INPUT_PATH = os.path.join(PARENT_FOLDER,'data/raw/')
RESULT_PATH = os.path.join(PARENT_FOLDER,'data/result/')

def read_data(path_name):
    '''
    Convert data to timestamp indexed dataframe
    :param path_name:
    :return:converted dataframe
    '''
    dateparse = lambda date: pd.datetime.strptime(date, "%Y-%m-%d %H:%M:%S")
    df = pd.read_csv(path_name, parse_dates=['Timestamp'], index_col='Timestamp', date_parser=dateparse)
    return df

def extract_features(df,weather):
    '''
    Extract features of input dataframe
    :param df: dataframe of current meter
    :param weather: weather of current meter
    :return: dataframe with extracted features
    '''
    df['month'] = df.index.month
    df['week'] = df.index.week
    df['hour'] = df.index.hour
    df['dayofyear'] = df.index.dayofyear
    df['dayofmonth'] = df.index.day
    df['dayofweek'] = df.index.weekday
    df['temperature'] = weather.loc[df.index].fillna(method='ffill')
    df['is_off'] = df.index.weekday > 4
    df['is_pre_off'] = (df.index.weekday == 0) |  (df.index.weekday == 6)
    df['is_next_off'] = (df.index.weekday == 4) | (df.index.weekday == 5)
    return df

def handle_holidays(df,holidays):
    for date in holidays['Date']:
        df.loc[date, 'is_off'] = True
        pre_date = str(datetime.strptime(date, '%Y-%m-%d').date() - timedelta(days=1))
        next_date = str(datetime.strptime(date, '%Y-%m-%d').date() + timedelta(days=1))
        df.loc[pre_date, 'is_next_off'] = True
        df.loc[next_date, 'is_pre_off'] = True
    df = df.dropna()
    df['is_off'] = df['is_off'].astype('bool')
    df['is_pre_off'] = df['is_pre_off'].astype('bool')
    df['is_next_off'] = df['is_next_off'].astype('bool')
    df.index = pd.to_datetime(df.index)
    return df


def xgb_model(df,n_fold):
    for i in range(n_fold):
        test_idx = list(range(int(len(df) * i / n_fold), int(len(df) * (i + 1) / n_fold)))
        train_idx = [x for x in range(len(df)) if x not in test_idx]
        features = ['month', 'week', 'hour', 'dayofyear',
                    'dayofmonth', 'dayofweek', 'temperature', 'is_off', 'is_pre_off',
                    'is_next_off']
        X_train = df.loc[:, features].iloc[train_idx, :]
        y_train = df.loc[:, 'Values'].iloc[train_idx]
        X_test = df.loc[:, features].iloc[test_idx, :]
        y_test = df.loc[:, 'Values'].iloc[test_idx]
        dtrain = xgb.DMatrix(X_train, y_train)
        dtest = xgb.DMatrix(X_test, y_test)
        # model training
        params = {'seed': 33,
                  'objective': 'reg:linear',
                  'silent': 0,
                  'nthread': 1,
                  'max_depth': 6,
                  'learning_rate': 0.3}
        num_round = 5000
        evallist = [(dtrain, 'train'), (dtest, 'eval')]
        bst = xgb.train(params, dtrain, num_round, evallist, early_stopping_rounds=100)
        df['pred' + str(i)] = bst.predict(xgb.DMatrix(df.loc[:, features]))
    return df

def get_anomalies(df,n):
    pred_cols = [col for col in df if col.startswith('pred')]
    df['error'] = df['Values'] - df[pred_cols].mean(1)
    df_day = df.resample('D').mean()
    anomalies = df_day.loc[df_day['error'] > df_day['error'].mean() + n * df_day['error'].std()]
    # anomalies = anomalies.loc[df.range > df.range.quantile(0.75)]
    return anomalies.index.astype('str').tolist()

def pipline(df,weather,holidays,n_fold,n_sigma):
    df = df.resample('H').mean().dropna()
    weather = weather.groupby(weather.index)['Temperature'].mean()
    df = extract_features(df, weather)
    if holidays is not None:
        df = handle_holidays(df, holidays)
    df = xgb_model(df, n_fold)
    anomalies = get_anomalies(df, n_sigma)
    return anomalies

def mark_anomalies(df,anomalies):
    '''
    mark abnormal dates of raw data
    :param df:
    :param anomalies:
    :return: df with abnormal result
    '''
    df['is_abnormal'] = False
    for date in anomalies:
        df.loc[date,'is_abnormal'] = True
    return df.reset_index()[['obs_id', 'meter_id','Timestamp','is_abnormal']]


if __name__=='__main__':
    #weather & holidays
    weather = read_data(INPUT_PATH+'weather.csv')
    holidays = pd.read_csv(INPUT_PATH+'holidays.csv').iloc[:,1:]

    # meter1, 234_203
    meter_id1 = '234_203'
    print('Detecting meter1 weekday anomalies...')
    df1 = read_data(DATA_PATH + meter_id1 + '.csv')
    weather_site1 = weather[(weather.site_id == meter_id1) & (weather['Distance'] > 10) & (weather['Distance'] < 13)].iloc[:,
    1:].fillna(method='ffill')[['Temperature']]
    # meter1 is a combination of two meters as we described in the report,
    # here we only cares the first part because it has more abnormal patterns with higher confidence
    anomalies1 = pipline(df1[:'2015-06-20'],weather_site1,None,3,2.3)
    res1 = mark_anomalies(df1,anomalies1)

    # meter2, 334_61
    meter_id2 = '334_61'
    print('Detecting meter2 weekday anomalies...')
    df2 = read_data(DATA_PATH + meter_id2 + '.csv')
    weather_site2 = weather[(weather.site_id == meter_id2) & (weather['Distance'] >= 19)].iloc[:, 1:].fillna(method='ffill')[
        ['Temperature']]
    holidays_site2 = holidays[holidays.site_id == meter_id2]
    anomalies2 = pipline(df2,weather_site2,holidays_site2,4,3.5)
    res2 = mark_anomalies(df2,anomalies2)

    # meter3, 38_9687
    meter_id3 = '38_9687'
    print('Detecting meter3 weekday anomalies...')
    df3 = read_data(DATA_PATH + meter_id3 + '.csv')
    weather_site3 = weather[weather.site_id == '38'].iloc[:, 1:].fillna(method='ffill')[['Temperature']]
    holidays_site3 = holidays[holidays.site_id == '038']
    # separate into two parts as there exists an obvious turning point on reactive meter curve
    anomalies3_part1 = pipline(df3[:'2012-12-21'],weather_site3,holidays_site3,3,3.5)
    anomalies3_part2 = pipline(df3['2012-12-21':],weather_site3,holidays_site3,4,3.5)
    anomalies3 = anomalies3_part1 + anomalies3_part2
    res3 = mark_anomalies(df3,anomalies3)

    # combine results of three meters
    print('Combining weekday result...')
    res = res1.append(res2).append(res3)
    res.to_csv(RESULT_PATH+'res_weekday_p.csv',index=False)



